contracts/KSElasticLMV2.sol

```diff
@@ -0,0 +1,1015 @@
+        fId = farmCount;
+        FarmInfo storage farm = farms[fId];
+
+        //validate phase input
+        _isPhaseValid(phase);
+
+        for (uint256 i; i < ranges.length; ) {
+            //validate range input
+            _isRangeValid(ranges[i]);
+
+            //push range into farm ranges array
+            farm.ranges.push(
+                IKSElasticLMV2.RangeInfo({
+                    tickLower: ranges[i].tickLower,
+                    tickUpper: ranges[i].tickUpper,
+                    weight: ranges[i].weight,
+                    isRemoved: false
+                })
+            );
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        //update farm data
+        farm.poolAddress = poolAddress;
+        farm.phase.startTime = phase.startTime;
+        farm.phase.endTime = phase.endTime;
+
+        for (uint256 i; i < phase.rewards.length; ) {
+            //push rewards info to farm phase rewards array
+            farm.phase.rewards.push(phase.rewards[i]);
+
+            //sumReward of newly created farm would be, this sumReward is total reward per liquidity until now
+            farm.sumRewardPerLiquidity.push(0);
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        //deploy farmingToken if needed
+        address destination;
+        if (isUsingToken) {
+            bytes memory creationCode = abi.encodePacked(
+                farmingTokenCreationCode,
+                abi.encode(msg.sender)
+            );
+            bytes32 salt = keccak256(abi.encode(msg.sender, fId));
+            assembly {
+                destination := create2(
+                    0,
+                    add(creationCode, 32),
+                    mload(creationCode),
+                    salt
+                )
+                if iszero(extcodesize(destination)) {
+                    revert(0, 0)
+                }
+            }
+            farm.farmingToken = destination;
+        }
+
+        //last touched time would be startTime
+        farm.lastTouchedTime = phase.startTime;
+
+        //increase farmCount
+        unchecked {
+            ++farmCount;
+        }
+
+        emit AddFarm(fId, poolAddress, ranges, phase, destination);
+    }
+
+    function addPhase(
+        uint256 fId,
+        PhaseInput calldata phaseInput
+    ) external isOperator {
+        if (fId >= farmCount) revert InvalidFarm();
+
+        //validate phase input
+        _isPhaseValid(phaseInput);
+
+        PhaseInfo storage phase = farms[fId].phase;
+
+        uint256 length = phase.rewards.length;
+        if (phaseInput.rewards.length != length) revert InvalidInput();
+
+        //if phase not settled, update sumReward.
+        //if phase already settled then it's not needed since sumReward would be unchanged
+        if (block.timestamp > farms[fId].lastTouchedTime && !phase.isSettled)
+            _updateFarmSumRewardPerLiquidity(fId);
```

Comment:
Should we move the check `block.timestamp > farms[fId].lastTouchedTime` into the `_updateFarmSumRewardPerLiquidity` function as it would be checked at any calls

Reply:
it would be check inside _calcSumRewardPerLiquidity.
In case block.timestamp == farms[fId].lastTouchedTime, _updateFarmSumRewardPerLiquidity would returns the lastest value.

Reply:
It's more consistent than we have to check it outside of each function that calls this function. Also we can ignore the result if we don't want to use

Reply:
updated, please help to check